src/bson_mapper/bson_streambuf.hpp

```diff
@@ -0,0 +1,78 @@
+// Copyright 2016 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/json.hpp>
+#include <mongocxx/collection.hpp>
+
+#include <bson_mapper/config/prelude.hpp>
+
+namespace bson_mapper {
+BSON_MAPPER_INLINE_NAMESPACE_BEGIN
+
+/**
+* A streambuffer that accepts bytes of BSON data,
+* and inserts a document into the given collection when
+* all bytes are sent over.
+*/
+class bson_output_streambuf : public std::streambuf {
```

Comment:
As I understand it, `final` would prevent another class from inheriting from `bson_output_streambuf`?

Could there be a situation in which one would want to inherit the class in order to do something fancier at the end of reading the BSON, instead of simply insert_one?